docs/Customization/MMIO-Peripherals.rst

```diff
@@ -0,0 +1,133 @@
+
+The ``LazyModule`` trait runs setup code that must execute before all the hardware gets elaborated.
+For a simple memory-mapped peripheral, this just involves connecting the peripheral's TileLink node to the MMIO crossbar.
+
+.. literalinclude:: ../../generators/example/src/main/scala/GCD.scala
+    :language: scala
+    :start-after: DOC include start: GCD lazy trait
+    :end-before: DOC include end: GCD lazy trait
+
+Note that the ``GCDTL`` class we created from the register router is itself a ``LazyModule``.
+Register routers have a TileLink node simply named "node", which we can hook up to the Rocket Chip bus.
+This will automatically add address map and device tree entries for the peripheral.
+Also observe how we have to place additional AXI4 buffers and converters for the AXI4 version of this peripheral.
+
+For peripherals which instantiate a concrete module, or which need to be connected to concrete IOs or wires, a matching concrete trait is necessary. We will make our GCD example output a ``gcd_busy`` signal as a top-level port to demonstrate. In the concrete module implementation trait, we instantiate the top level IO (a concrete object) and wire it to the IO of our lazy module.
+
+
+.. literalinclude:: ../../generators/example/src/main/scala/GCD.scala
+    :language: scala
+    :start-after: DOC include start: GCD imp trait
+    :end-before: DOC include end: GCD imp trait
+
+Constructing the Top and Config
+-------------------------------
+
+Now we want to mix our traits into the system as a whole.
+This code is from ``generators/example/src/main/scala/Top.scala``.
+
+.. literalinclude:: ../../generators/example/src/main/scala/Top.scala
+    :language: scala
+    :start-after: DOC include start: Top
+    :end-before: DOC include end: Top
+
+Just as we need separate traits for ``LazyModule`` and module implementation, we need two classes to build the system.
+The ``Top`` class contains the set of traits which parameterize and define the ``Top``. Typically these traits will optionally add IOs or peripherals to the ``Top``.
+The ``Top`` class includes the pre-elaboration code and also a ``lazy val`` to produce the module implementation (hence ``LazyModule``).
+The ``TopModule`` class is the actual RTL that gets synthesized.
+
+And finally, we create a configuration class in ``generators/example/src/main/scala/Configs.scala`` that uses this mixin.
+
+.. literalinclude:: ../../generators/example/src/main/scala/RocketConfigs.scala
+    :language: scala
+    :start-after: DOC include start: GCDTLRocketConfig
+    :end-before: DOC include end: GCDTLRocketConfig
+
+Testing
+-------
+
+Now we can test that the GCD is working. The test program is in ``tests/gcd.c``.
+
+.. literalinclude:: ../../tests/gcd.c
+    :language: c
+
+This just writes out to the registers we defined earlier.
+The base of the module's MMIO region is at 0x2000 by default.
+This will be printed out in the address map portion when you generated the verilog code.
```

Comment:
I think this would be a good place to mention that there are `json` files that have the address maps.